### PR TITLE
[MGX-332] Automatic fee unlock

### PR DESCRIPTION
**Pull request by mateuszaaa**
- change hashing algorithm so account ids from  are fetchable
- Implement on_idle that automatically unlocks outdated fee locks
- unit tests for on idle hook

### CODE REVIEW

**Review by vanderian — DISMISSED** [review]

File: pallets/fee-lock/src/tests.rs [hunk 1]
```diff
@@ -720,3 +721,194 @@ fn whitelist_and_valuation_works() {
...
+fn test_on_idle_unlock_for_single_user(
+	availabe_weight: Weight,
+	consumed_weight: Weight,
+	expected_account_data: AccountData<Balance>,
+) {
+	ExtBuilder::new()
+		.create_token(NativeCurrencyId::get())
+		.mint(ALICE, NativeCurrencyId::get(), INITIAL_AMOUNT)
+		.initialize_fee_locks(PERIOD_LENGTH, FEE_LOCK_AMOUNT, SWAP_VALUE_THRESHOLD)
+		.build()
```
Comment (mateuszaaa) [inline pallets/fee-lock/src/tests.rs:786]: some nice way i found for setuping test environemnt ;)

File: pallets/fee-lock/src/lib.rs [hunk 2]
```diff
@@ -60,7 +60,50 @@ pub mod pallet {
...
+						if matches!(unlock_block, Some(unlock) if unlock <= now) {
+							consumed_weight += T::WeightInfo::unlock_fee();
+							let _ = <Self as FeeLockTriggerTrait<T::AccountId>>::unlock_fee(&who);
+						}
+					},
+					_ => break,
+				};
+
+				if cost_of_single_unlock_iteration.ref_time() >
+					(remaining_weight.ref_time() - consumed_weight.ref_time())
```
[the remark this answers is not in the document]
Comment (majusko) [reply pallets/fee-lock/src/lib.rs:99]: Agree, we should probably at least add a warning log on all `break` cases.

[the remark this answers is not in the document]
Comment (mateuszaaa) [reply pallets/fee-lock/src/lib.rs:124]: > Feels a bit risky. I understand it will work correctly with the current impl but any modifications later on will make it panic. 

What exactly is risky here, actually i dont think its possible to make to working wrong, if accounts for every manual storage read/write & uses the bench marked weights for extrinsics.

So to make it not work you would have to modify on_idle hook itself, then its your responsibility to take care of things,  i also came up with a lot of unit tests for all the important corner cases i guess

```
running 20 tests
test mock::__construct_runtime_integrity_test::runtime_integrity_tests ... ok
test tests::test_on_idle_unlock_multiple_users::unlocks_tokens_for_single_account_only_with_limited_weight ... ok
test tests::test_autounlock_on_empty_unlock_queue ... ok
test tests::test_on_idle_unlock_for_single_user::does_not_unlock_tokens_when_weigh_is_zero ... ok
test tests::test_on_idle_unlock_for_single_user::unlock_tokens_using_a_bit_more_weight_that_required ... ok
test tests::test_on_idle_unlock_for_single_user::unlocks_tokens_for_an_user ... ok
test tests::test_unlock_happens_not_sooner_but_after_period ... ok
test tests::test_on_idle_unlock_multiple_users::unlocks_tokens_for_both_users_using_exact_required_weight_ ... ok
test tests::test_on_idle_unlock_multiple_users::unlocks_tokens_for_both_users_with_unlimited_input_weight ... ok
test tests::test_maintain_queue_with_subsequent_fee_locks_on_single_account ... ok
test tests::test_process_queue_and_ignore_outdated_items_in_unlock_queue_because_of_subsequent_process_fee_lock_calls ... ok
test tests::process_fee_lock_trigger_works ... ok
test tests::test_process_queue_and_ignore_outdated_items_in_unlock_queue_because_of_manual_unlock ... ok
test tests::test_unlock_happens_in_order ... ok
test tests::test_on_idle_unlock_for_single_user::unlock_tokens_using_a_too_small_weight_that_required ... ok
test tests::test_on_idle_unlock_for_single_user::unlock_tokens_using_exact_amount_of_weight_required ... ok
test tests::test_unlock_stops_after_single_iteration_without_consuming_unnecessary_weight ... ok
test tests::update_fee_lock_metadata_works ... ok
test tests::whitelist_and_valuation_works ... ok
test tests::unlock_fee_works ... ok

```

[the remark this answers is not in the document]
Comment (iStrike7) [reply pallets/fee-lock/src/lib.rs:123]: If the tests cover it and we are not modifying the hook without due considerations, then its fine

**Review by majusko — DISMISSED** [review]
